cpp/tests/test_allgather.cpp

```diff
@@ -0,0 +1,289 @@
+    EXPECT_EQ(metadata.size(), n_elements * sizeof(int));
+
+    for (int i = 0; i < n_elements; i++) {
+        int val;
+        std::memcpy(&val, metadata.data() + i * sizeof(int), sizeof(int));
+        EXPECT_EQ(offset + i, val);
+    }
+
+    EXPECT_EQ(packed_data.data->size, n_elements * sizeof(int));
+    std::vector<uint8_t> copied_data(n_elements * sizeof(int));
+    RAPIDSMPF_CUDA_TRY(cudaMemcpyAsync(
+        copied_data.data(),
+        packed_data.data->data(),
+        n_elements * sizeof(int),
+        cudaMemcpyDefault,
+        stream
+    ));
+
+    RAPIDSMPF_CUDA_TRY(cudaStreamSynchronize(stream));
+    EXPECT_EQ(metadata, copied_data);
+}
+
+class BaseAllGatherTest : public ::testing::Test {
+  protected:
+    void SetUp() override {
+        GlobalEnvironment->barrier();
+
+        stream = cudf::get_default_stream();
+        mr = std::make_unique<rmm::mr::cuda_memory_resource>();
+        br = std::make_unique<rapidsmpf::BufferResource>(mr.get());
+        comm = GlobalEnvironment->comm_.get();
+
+        allgather = std::make_unique<AllGather>(
+            GlobalEnvironment->comm_,
+            GlobalEnvironment->progress_thread_,
+            0,
+            stream,
+            br.get()
+        );
+    }
+
+    void TearDown() override {
+        allgather = nullptr;
+        br = nullptr;
+        mr = nullptr;
+        GlobalEnvironment->barrier();
+    }
+
+    rmm::cuda_stream_view stream;
+    rapidsmpf::Communicator* comm;
+    std::unique_ptr<rapidsmpf::BufferResource> br;
+    std::unique_ptr<AllGather> allgather;
+    std::unique_ptr<rmm::mr::device_memory_resource> mr;
+};
+
+// Test simple shutdown
+TEST_F(BaseAllGatherTest, shutdown) {}
+
+class AllGatherTest : public BaseAllGatherTest,
+                      public ::testing::WithParamInterface<std::tuple<int, int, bool>> {
+  protected:
+    void SetUp() override {
+        BaseAllGatherTest::SetUp();
+        std::tie(n_elements, n_inserts, ordered) = GetParam();
+    }
+
+    int n_elements;
+    int n_inserts;
+    bool ordered;
+};
+
+// Parameterized test for different element counts
+INSTANTIATE_TEST_SUITE_P(
+    AllGather,
+    AllGatherTest,
+    ::testing::Combine(
+        ::testing::Values(0, 1, 10, 100),  // n_elements
+        ::testing::Values(0, 1, 10),  // n_inserts
+        ::testing::Values(false, true)  // ordered
+    ),
+    [](const ::testing::TestParamInfo<AllGatherTest::ParamType>& info) {
+        return "n_elements_" + std::to_string(std::get<0>(info.param)) + "_n_inserts_"
+               + std::to_string(std::get<1>(info.param)) + "_"
+               + (std::get<2>(info.param) ? "ordered" : "unordered");
+    }
+);
+
+constexpr auto gen_offset(int i, int r) {
+    return i * 10 + r;
+};
+
+TEST_P(AllGatherTest, basic_allgather) {
+    auto this_rank = comm->rank();
+
+    for (int i = 0; i < n_inserts; i++) {
+        auto packed_data =
+            generate_packed_data(n_elements, gen_offset(i, this_rank), stream, *br);
```

Comment:
Can we use a different number of elements on each rank, to test the `allgather_v_`-ness of this.